```
src/index/unittest/test_nsg/CMakeFiles/test_nsg.dir/build.make:75: recipe for target 'src/index/unittest/test_nsg/CMakeFiles/test_nsg.dir/__/__/knowhere/knowhere/index/vector_index/IndexNSG.cpp.o' failed
make[2]: *** [src/index/unittest/test_nsg/CMakeFiles/test_nsg.dir/__/__/knowhere/knowhere/index/vector_index/IndexNSG.cpp.o] Error 1
CMakeFiles/Makefile2:1132: recipe for target 'src/index/unittest/test_nsg/CMakeFiles/test_nsg.dir/all' failed
make[1]: *** [src/index/unittest/test_nsg/CMakeFiles/test_nsg.dir/all] Error 2
```

--- a/core/src/db/engine/ExecutionEngineImpl.cpp
+++ b/core/src/db/engine/ExecutionEngineImpl.cpp
@@ -611,6 +611,9 @@ ExecutionEngineImpl::Init() {
     server::Config& config = server::Config::GetInstance();
     std::vector<int64_t> gpu_ids;
     Status s = config.GetGpuResourceConfigBuildIndexResources(gpu_ids);
+    if (!s.ok()) {
+        gpu_num_ = knowhere::INVALID_VALUE;
+    }
     for (auto id : gpu_ids) {
         if (gpu_num_ == id) {
             return Status::OK();

--- a/core/src/grpc/README.md
+++ b/core/src/grpc/README.md
@@ -1,6 +1,0 @@
-We manually change two APIs in "milvus.pb.h":
-    add_vector_data()
-    add_row_id_array()
-    add_ids()
-    add_distances()
-If proto files need be generated again, remember to re-change above APIs.

--- a/core/src/index/knowhere/knowhere/index/vector_index/IndexNSG.cpp
+++ b/core/src/index/knowhere/knowhere/index/vector_index/IndexNSG.cpp
@@ -116,29 +116,29 @@ NSG::Train(const DatasetPtr& dataset, const Config& config) {
     }
 
     // TODO(linxj): dev IndexFactory, support more IndexType
-    bool use_gpu = false;
-#ifdef MILVUS_GPU_VERSION
-    use_gpu = true;
-    auto temp_resource = FaissGpuResourceMgr::GetInstance().GetRes(build_cfg->gpu_id);
-    if (temp_resource == nullptr)
-        use_gpu = false;
-#endif
     Graph knng;
-    if (use_gpu) {
-        auto preprocess_index = std::make_shared<GPUIVF>(build_cfg->gpu_id);
-        auto model = preprocess_index->Train(dataset, config);
-        preprocess_index->set_index_model(model);
-        preprocess_index->AddWithoutIds(dataset, config);
-
-        preprocess_index->GenGraph(build_cfg->knng, knng, dataset, config);
-    } else {
+#ifdef MILVUS_GPU_VERSION
+    auto temp_resource = FaissGpuResourceMgr::GetInstance().GetRes(build_cfg->gpu_id);
+    if (temp_resource == nullptr) {
         auto preprocess_index = std::make_shared<IVF>();
         auto model = preprocess_index->Train(dataset, config);
         preprocess_index->set_index_model(model);
         preprocess_index->AddWithoutIds(dataset, config);
-
+        preprocess_index->GenGraph(build_cfg->knng, knng, dataset, config);
+    } else {
+        auto preprocess_index = std::make_shared<GPUIVF>(build_cfg->gpu_id);
+        auto model = preprocess_index->Train(dataset, config);
+        preprocess_index->set_index_model(model);
+        preprocess_index->AddWithoutIds(dataset, config);
         preprocess_index->GenGraph(build_cfg->knng, knng, dataset, config);
     }
+#else
+    auto preprocess_index = std::make_shared<IVF>();
+    auto model = preprocess_index->Train(dataset, config);
+    preprocess_index->set_index_model(model);
+    preprocess_index->AddWithoutIds(dataset, config);
+    preprocess_index->GenGraph(build_cfg->knng, knng, dataset, config);
+#endif
 
     algo::BuildParams b_params;
     b_params.candidate_pool_size = build_cfg->candidate_pool_size;

--- a/core/src/index/knowhere/knowhere/index/vector_index/IndexNSG.cpp
+++ b/core/src/index/knowhere/knowhere/index/vector_index/IndexNSG.cpp
@@ -118,8 +118,7 @@ NSG::Train(const DatasetPtr& dataset, const Config& config) {
     // TODO(linxj): dev IndexFactory, support more IndexType
     Graph knng;
 #ifdef MILVUS_GPU_VERSION
-    auto temp_resource = FaissGpuResourceMgr::GetInstance().GetRes(build_cfg->gpu_id);
-    if (temp_resource == nullptr) {
+    if (build_cfg->gpu_id == knowhere::INVALID_VALUE) {
         auto preprocess_index = std::make_shared<IVF>();
         auto model = preprocess_index->Train(dataset, config);
         preprocess_index->set_index_model(model);

--- a/core/src/wrapper/ConfAdapter.cpp
+++ b/core/src/wrapper/ConfAdapter.cpp
@@ -39,8 +39,6 @@ void
 ConfAdapter::MatchBase(knowhere::Config conf) {
     if (conf->metric_type == knowhere::DEFAULT_TYPE)
         conf->metric_type = knowhere::METRICTYPE::L2;
-    if (conf->gpu_id == knowhere::INVALID_VALUE)
-        conf->gpu_id = 0;
 }
 
 knowhere::Config

